plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -1846,6 +1868,61 @@ private void prepareAlterPartition(HdfsContext hdfsContext, HiveIdentity identit
                     new PartitionWithStatistics(oldPartition.get(), partitionName, oldPartitionStatistics)));
         }
 
+        private void handleFilesForFailedRetryAttempts(HdfsContext hdfsContext, String queryId, PartitionAndMore partitionAndMore)
+        {
+            if (!partitionAndMore.isRetriesEnabled()) {
+                return;
+            }
+            if (!partitionAndMore.hasFileNames()) {
+                return;
+            }
+            deleteFilesForFailedRetryAttempts(hdfsContext, queryId, partitionAndMore.getCurrentLocation(), ImmutableSet.copyOf(partitionAndMore.getFileNames()));
+        }
+
+        private void handleFilesForFailedRetryAttempts(HdfsContext hdfsContext, String queryId, TableAndMore tableAndMore)
+        {
+            if (!tableAndMore.isRetriesEnabled()) {
+                return;
+            }
+            if (tableAndMore.getCurrentLocation().isEmpty()) {
+                return;
+            }
+            if (tableAndMore.getFileNames().isEmpty()) {
+                return;
+            }
+
+            Path tableLocation = tableAndMore.getCurrentLocation().orElseThrow(() -> new IllegalArgumentException("table location not set"));
+            List<String> files = tableAndMore.getFileNames().orElseThrow(() -> new IllegalStateException("file list not not set"));
+            deleteFilesForFailedRetryAttempts(hdfsContext, queryId, tableLocation, ImmutableSet.copyOf(files));
+        }
+
+        private void deleteFilesForFailedRetryAttempts(HdfsContext hdfsContext, String queryId, Path path, Set<String> files)
+        {
+            ImmutableList.Builder<String> deletedFilesBuilder = ImmutableList.builder();
+            try {
+                FileSystem fileSystem = hdfsEnvironment.getFileSystem(hdfsContext, path);
+                if (!fileSystem.exists(path)) {
+                    return;
+                }
+                RemoteIterator<LocatedFileStatus> iterator = fileSystem.listFiles(path, false);
+                while (iterator.hasNext()) {
+                    Path file = iterator.next().getPath();
+                    if (isFileCreatedByQuery(file.getName(), queryId) && !files.contains(file.getName())) {
+                        fileSystem.delete(file, false);
+                        deletedFilesBuilder.add(file.getName());
+                    }
+                }
+
+                List<String> deleteFiles = deletedFilesBuilder.build();
+                if (!deleteFiles.isEmpty()) {
+                    log.info("Deleted failed attempt files %s from %s for query %s", deleteFiles, path, queryId);
+                }
+            }
+            catch (IOException e) {
+                throw new TrinoException(HIVE_FILESYSTEM_ERROR, format("Error removing failed retry attempt files from %s", path), e);
```

Comment:
If we fail here query will be rolled back. The optimal outcome would be for rollback to complete successfully and clean up everything for query. 
Yet if we have problem here, probably rollback will also fail.

What we can do is to list files that we are trying to delete here in an exception we are throwing from here. So those can be cleaned up later by user manually.
Still it is not great. The rollback routine would still fire and try to cleanup the changes query made. It will cleanup some, leave some behind probably. It is not obvious that if at this 
point user cleans up the failed attempt files the table would be in the expected state.

tl;dr; if we are using non-transactional connector like Hive we are screwed in case of the failure here. At least I do not see a great solution.

Reply:
Added comment. And also changed the code to have a list of remaining files to be deleted in error message.